File: py/cd_inversion/cd_forward_model.py
"""CD spectroscopy forward model."""
import numpy as np

WAVELENGTHS = np.arange(190.0, 251.0, 1.0)  # 61 points
N_WL = len(WAVELENGTHS)

def _build_basis():
    wl = WAVELENGTHS
    helix = (-30*np.exp(-0.5*((wl-208)/5)**2) - 25*np.exp(-0.5*((wl-222)/4)**2)
             + 50*np.exp(-0.5*((wl-193)/8)**2))
    sheet = (-15*np.exp(-0.5*((wl-216)/7)**2) + 20*np.exp(-0.5*((wl-195)/6)**2))
    coil  = (-8*np.exp(-0.5*((wl-200)/6)**2) + 5*np.exp(-0.5*((wl-218)/8)**2))
    return helix, sheet, coil

HELIX_BASIS, SHEET_BASIS, COIL_BASIS = _build_basis()
REFERENCE_SPECTRA = {"helix": HELIX_BASIS, "sheet": SHEET_BASIS, "coil": COIL_BASIS}


def predict_spectrum(h: float, s: float, c: float) -> np.ndarray:
    tot = h + s + c
    if tot < 1e-6:
        h = s = c = 1/3; tot = 1.0
    return (h/tot)*HELIX_BASIS + (s/tot)*SHEET_BASIS + (c/tot)*COIL_BASIS


def add_noise(spectrum: np.ndarray, sigma: float = 0.5, seed: int = 0) -> np.ndarray:
    rng = np.random.default_rng(seed)
    return spectrum + sigma * rng.standard_normal(len(spectrum))

File: jl/cd_inversion/src/CDForwardModel.jl
module CDForwardModel

export CDParams, REFERENCE_SPECTRA, predict_spectrum, add_noise

# Wavelength range: 190-250 nm at 1 nm steps
const WAVELENGTHS = collect(190.0:1.0:250.0)   # 61 points

Base.@kwdef struct CDParams
    wavelengths::Vector{Float64} = WAVELENGTHS
    noise_level::Float64 = 0.01
end

# Reference CD basis spectra from literature (millidegrees per nm)
# Based on classical CD spectroscopy for secondary structure
const REFERENCE_SPECTRA = let
    λ = WAVELENGTHS
    n = length(λ)

    # α-helix: strong negative at 208 and 222 nm, positive at 193 nm
    helix = zeros(n)
    for (i, wl) in enumerate(λ)
        helix[i]  = -30.0 * exp(-0.5*((wl-208)/5)^2)
        helix[i] += -25.0 * exp(-0.5*((wl-222)/4)^2)
        helix[i] +=  50.0 * exp(-0.5*((wl-193)/8)^2)
    end

    # β-sheet: negative around 216 nm, positive around 195 nm
    sheet = zeros(n)
    for (i, wl) in enumerate(λ)
        sheet[i]  = -15.0 * exp(-0.5*((wl-216)/7)^2)
        sheet[i] +=  20.0 * exp(-0.5*((wl-195)/6)^2)
    end

    # Random coil: weak negative around 200 nm, positive around 218 nm
    coil = zeros(n)
    for (i, wl) in enumerate(λ)
        coil[i]  = -8.0 * exp(-0.5*((wl-200)/6)^2)
        coil[i] +=  5.0 * exp(-0.5*((wl-218)/8)^2)
    end

    Dict("helix" => helix, "sheet" => sheet, "coil" => coil)
end

# Linear mixture model: CD = f_h * S_h + f_s * S_s + f_c * S_c
function predict_spectrum(helix_frac::Float64, sheet_frac::Float64, coil_frac::Float64,
                          params::CDParams=CDParams())::Vector{Float64}
    h = clamp(helix_frac, 0.0, 1.0)
    s = clamp(sheet_frac, 0.0, 1.0)
    c = clamp(coil_frac,  0.0, 1.0)
    # Normalise so fractions sum to 1
    tot = h + s + c
    tot < 1e-6 && (h = 1.0/3; s = 1.0/3; c = 1.0/3; tot = 1.0)
    h /= tot; s /= tot; c /= tot
    h .* REFERENCE_SPECTRA["helix"] .+ s .* REFERENCE_SPECTRA["sheet"] .+ c .* REFERENCE_SPECTRA["coil"]
end

function add_noise(spectrum::Vector{Float64}; σ::Float64=0.5, seed::Int=0)::Vector{Float64}
    rng = Random.MersenneTwister(seed)
    spectrum .+ σ .* randn(rng, length(spectrum))
end

end
